<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Project-Overview" data-toc-modified-id="Project-Overview-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Project Overview</a></span></li><li><span><a href="#Data-Scraping-Specifics" data-toc-modified-id="Data-Scraping-Specifics-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data Scraping Specifics</a></span></li><li><span><a href="#Objectives" data-toc-modified-id="Objectives-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Objectives</a></span></li><li><span><a href="#Initial-Extraction-of-Data" data-toc-modified-id="Initial-Extraction-of-Data-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Initial Extraction of Data</a></span><ul class="toc-item"><li><span><a href="#Imports-and-Settings" data-toc-modified-id="Imports-and-Settings-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Imports and Settings</a></span></li><li><span><a href="#Quick-Overview" data-toc-modified-id="Quick-Overview-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Quick Overview</a></span></li><li><span><a href="#Data-Extraction-from-FlexTable_info" data-toc-modified-id="Data-Extraction-from-FlexTable_info-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Data Extraction from FlexTable_info</a></span></li><li><span><a href="#Data-Extraction-from-FeaturesTable_info" data-toc-modified-id="Data-Extraction-from-FeaturesTable_info-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Data Extraction from FeaturesTable_info</a></span></li><li><span><a href="#Data-Extraction-from-MainTable_info" data-toc-modified-id="Data-Extraction-from-MainTable_info-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Data Extraction from MainTable_info</a></span></li><li><span><a href="#Exploding-Data-Frame-Entries-to-Get-Sub-Listings" data-toc-modified-id="Exploding-Data-Frame-Entries-to-Get-Sub-Listings-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>Exploding Data Frame Entries to Get Sub-Listings</a></span></li></ul></li><li><span><a href="#Cleaning-Data-in-Sub-Categories" data-toc-modified-id="Cleaning-Data-in-Sub-Categories-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Cleaning Data in Sub-Categories</a></span><ul class="toc-item"><li><span><a href="#Cleaning-Data-in-MainTable_info" data-toc-modified-id="Cleaning-Data-in-MainTable_info-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Cleaning Data in MainTable_info</a></span><ul class="toc-item"><li><span><a href="#Cleaning-Listing_Type-Column" data-toc-modified-id="Cleaning-Listing_Type-Column-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Cleaning Listing_Type Column</a></span></li><li><span><a href="#Cleaning-Sq_Feet-Column" data-toc-modified-id="Cleaning-Sq_Feet-Column-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>Cleaning Sq_Feet Column</a></span></li><li><span><a href="#Cleaning-Furnishing-Column" data-toc-modified-id="Cleaning-Furnishing-Column-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>Cleaning Furnishing Column</a></span></li><li><span><a href="#Cleaning-Baths-Column" data-toc-modified-id="Cleaning-Baths-Column-5.1.4"><span class="toc-item-num">5.1.4&nbsp;&nbsp;</span>Cleaning Baths Column</a></span></li><li><span><a href="#Cleaning-Beds-Column" data-toc-modified-id="Cleaning-Beds-Column-5.1.5"><span class="toc-item-num">5.1.5&nbsp;&nbsp;</span>Cleaning Beds Column</a></span></li><li><span><a href="#Cleaning-Rent_Monthly-Column" data-toc-modified-id="Cleaning-Rent_Monthly-Column-5.1.6"><span class="toc-item-num">5.1.6&nbsp;&nbsp;</span>Cleaning Rent_Monthly Column</a></span></li><li><span><a href="#Cleaning-Utilities_info-Column" data-toc-modified-id="Cleaning-Utilities_info-Column-5.1.7"><span class="toc-item-num">5.1.7&nbsp;&nbsp;</span>Cleaning Utilities_info Column</a></span></li></ul></li><li><span><a href="#Cleaning-Data-in-FlexTable_info" data-toc-modified-id="Cleaning-Data-in-FlexTable_info-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Cleaning Data in FlexTable_info</a></span><ul class="toc-item"><li><span><a href="#Cleaning-Neighbourhood-Column" data-toc-modified-id="Cleaning-Neighbourhood-Column-5.2.1"><span class="toc-item-num">5.2.1&nbsp;&nbsp;</span>Cleaning Neighbourhood Column</a></span></li><li><span><a href="#Cleaning-Smoking-Column" data-toc-modified-id="Cleaning-Smoking-Column-5.2.2"><span class="toc-item-num">5.2.2&nbsp;&nbsp;</span>Cleaning Smoking Column</a></span></li><li><span><a href="#Cleaning-Landlord_On_Rentfaster_Since-Column" data-toc-modified-id="Cleaning-Landlord_On_Rentfaster_Since-Column-5.2.3"><span class="toc-item-num">5.2.3&nbsp;&nbsp;</span>Cleaning Landlord_On_Rentfaster_Since Column</a></span></li><li><span><a href="#Cleaning-Year_Built-Column" data-toc-modified-id="Cleaning-Year_Built-Column-5.2.4"><span class="toc-item-num">5.2.4&nbsp;&nbsp;</span>Cleaning Year_Built Column</a></span></li><li><span><a href="#Cleaning-Pets-Column" data-toc-modified-id="Cleaning-Pets-Column-5.2.5"><span class="toc-item-num">5.2.5&nbsp;&nbsp;</span>Cleaning Pets Column</a></span></li><li><span><a href="#Cleaning-Tenant_Parking-Column" data-toc-modified-id="Cleaning-Tenant_Parking-Column-5.2.6"><span class="toc-item-num">5.2.6&nbsp;&nbsp;</span>Cleaning Tenant_Parking Column</a></span></li></ul></li><li><span><a href="#Cleaning-Data-in-FeaturesTable_info" data-toc-modified-id="Cleaning-Data-in-FeaturesTable_info-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Cleaning Data in FeaturesTable_info</a></span><ul class="toc-item"><li><span><a href="#Cleaning-Building_features-Column" data-toc-modified-id="Cleaning-Building_features-Column-5.3.1"><span class="toc-item-num">5.3.1&nbsp;&nbsp;</span>Cleaning Building_features Column</a></span></li><li><span><a href="#Cleaning-Community_features-Column" data-toc-modified-id="Cleaning-Community_features-Column-5.3.2"><span class="toc-item-num">5.3.2&nbsp;&nbsp;</span>Cleaning Community_features Column</a></span></li><li><span><a href="#Cleaning-Property_features-Column" data-toc-modified-id="Cleaning-Property_features-Column-5.3.3"><span class="toc-item-num">5.3.3&nbsp;&nbsp;</span>Cleaning Property_features Column</a></span></li></ul></li></ul></li><li><span><a href="#Removing-Repeated-Listings" data-toc-modified-id="Removing-Repeated-Listings-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Removing Repeated Listings</a></span><ul class="toc-item"><li><span><a href="#Understanding-the-Issue" data-toc-modified-id="Understanding-the-Issue-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Understanding the Issue</a></span></li><li><span><a href="#Getting-Number-of-Sub-Listings-per-Listing" data-toc-modified-id="Getting-Number-of-Sub-Listings-per-Listing-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Getting Number of Sub-Listings per Listing</a></span></li><li><span><a href="#Getting-Index-at-Repetitions-and-Cleaning" data-toc-modified-id="Getting-Index-at-Repetitions-and-Cleaning-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Getting Index at Repetitions and Cleaning</a></span></li></ul></li><li><span><a href="#Saving-Final-Dataframe" data-toc-modified-id="Saving-Final-Dataframe-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Saving Final Dataframe</a></span></li><li><span><a href="#Next-Steps" data-toc-modified-id="Next-Steps-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Next Steps</a></span></li></ul></div>

# Project Overview


I wanted to leverage data science to build a tool that would predict the montlhy rent of rental listings in Calgary depending on the different features of the property. Such a tool would be benifical for many people like:

> 1. __New Immigrants__ : Many immigrants are settling in Calgary and they would be interested in calculating their monthly expenditure here so that they can better manage their finances. 
> 2. __International Students__ : There are many international students studying at places like University of Calgary and they would want a baseline to negotiate their rent against.
> 3. __Landlords__ : Landlords can gauge what is the general asking price for a similar listing as theirs and then put up a more competitive monthly rent to attract renters.



In order to gather data for analysis, I scraped the rental listings from <a href="https://www.rentfaster.ca/" target="_blank">RentFaster</a>, which is one of the most popular rental listings platform in Canada. I used Selenium library to scrape ~1500 listings from the website for listings in Calgary and stored the raw data in CSV format (<a href="https://github.com/RahulNagial/Regression_Analysis" target="_blank">refer "rentfaster_scraped_data.csv"</a>). For the scraping script used, please <a href="https://github.com/RahulNagial/Regression_Analysis/blob/main/rentfasterscraping.py" target="_blank">refer "rentfasterscraping.py"</a>.

In the present and subsequent notebooks, I cleaned the scraped data, performed exploratory data analysis on it and used the insights gained to build machine learning models on it (i.e. regression models) to predict montlhy rent. I tuned the model and got the best model with a mean absolute error as $160. Finally, I productionized the best model using Flask API and rendered a client-facing UI using HTML. Finally, I hosted it on a local webserver. <a href="https://github.com/RahulNagial/Regression_Analysis" target="_blank">For more on this project refer my GitHub repository</a>.


# Data Scraping Specifics

It is important to understand the structure of the website from where data was scraped. This understanding would help in guiding the workflow for cleaning scraped data in this notebook. 

The geckodriver for Mozilla Firefox used in Selenium scraping script followed the following protocol. First, it openend the browser window and went to the <a href="https://www.rentfaster.ca/ab/calgary/rentals/#dialog-listview" target="_blank">url</a> containing all listing for rentals in Calgary. The landing page looked as shown below. There were multiple listings on each page and there were ~120 pages comprising all the rental listings in Calgary. The geckodriver clicked on each individual listing (highlighted in red) on the page, scraped the data, closed the individual listing and then went on to next one. 

Once the end of the page was reached, the gecko driver clicked on the 'Next' button (highlighted in red) to go to next page and this process continued till ~1500 listings were scraped. 

 ![landingpage](images/landingpage.png)

When each of the the individual listing on a page was clicked on, the window shown below popped up and it contained all the required information for scraping purposes. The information was contained in the following sub-categories of the pop-up (as highlighted in red):

> 1. __Flex Table__ 
> 2. __Main Info__
> 3. __Full Description__
> 4. __Features Table__


 ![Rentfaster Image](images/broadstructure.png)

An important thing to notice in Main Info section is that each individual listing can have many sub-listings and thus we have to handle these sub-listings by expanding them (i.e. *exploding* them latter).

Given the structure of rentfaster website, the data is extracted and grouped in the above mentioned four major sub-categories in the raw scraped file. The aim next is to unpack the raw data and extract information from it.

# Objectives

 The objectives of this notebook are to extract the following information from raw data:
 
> 1. __Listing type (from Main Info)__
> 2. __Size of listing in sq. feet (from Main Info)__
> 3. __Whether it is furnished or not (from Main Info)__
> 4. __Number of bathrooms (from Main Info)__
> 5. __Number of bedrooms (from Main Info)__
> 6. __Monthly rent amount (from Main Info)__
> 7. __How many utilities are included (from Main Info)__
> 8. __Neighbourhood name for listing (from Flex Table)__
> 9. __Is smoking allowed in listing? (from Flex Table)__
> 10. __Number of years landlord has been on Rentfaster (from Flex Table)__
> 11. __Year when property was built (from Flex Table)__
> 12. __Whether pets are allowed or not? (from Flex Table)__
> 13. __Is tenant parking covered and protected? (from Flex Table)__
> 14. __What are the features meantioned for building, property and community? Also, ranking their features and using the building/property/community score for subsequent analysis. (from Features Table)__


In more general terms, the objective of this notebook is to provide a walk-through of the data cleaning process for the raw scraped data. 

# Initial Extraction of Data

The data cleaning work starts in this sub-section. 

## Imports and Settings

First, I import the required libraries and setup the notebook enviornment. I also load the raw data in *df*.

In [1]:
# importing necessary libraries
import itertools
import pandas as pd
import numpy as np

# turning off warnings for the notebook
import warnings
warnings.filterwarnings('ignore')

# displaying all the columns and row throughout the notebook
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)



In [2]:
# reading in the raw data and creating a copy of it to work on
df_rentfaster = pd.read_csv("rentfaster_scraped_data.csv")
df = df_rentfaster.copy()

## Quick Overview

I checked a few enteries and shape of the dataframe.

In [3]:
# quick look at the data
df_rentfaster[:20]
# df.tail()

Listings_Title  \
0                                       VISTA PHASE 1   
1   LARGE 2 BEDROOMS WITH DEN CONDO AT TARALAKE WA...   
2   COZY 2 BEDROOM COTTAGE HOUSE WITH HUGE FENCED ...   
3                           VARSITY TOWERS NORTH WEST   
4                                       VISTA PHASE 2   
5                                       VISTA PHASE 2   
6                                       VISTA PHASE 1   
7              6 BED,LIVING ROOM,FAMILY ROOM,3.5 BATH   
8         2 BED/2 BATH CONTEMPORARY CONDO IN PARKHILL   
9   MOVE IN NOW AND DON'T PAY UNTIL MAY 1 + GET AP...   
10                                          APARTMENT   
11              5 BEDROOM INNERCITY HOME ON 17 AVE SW   
12   SPRUCE RIDGE ESTATES (104-19A HEMLOCK CRES. SW )   
13   PATRICIAN VILLAGE (L208, 1919 UNIVERSITY DR. NW)   
14                                     LOFTS ON FIRST   
15  BEAUTIFUL 2 BDRM BUNGALOW TOWNHOUSE - STEPS FR...   
16                    BRIGHT AND COZY 1 BEDROOM SUITE   
17                                    WESTBROOK MANOR   
18        BEAUTIFUL 3 BEDROOM SUITE NEAR UOFC AND LRT   
19                                MAINSTREET ON ELBOW   

                            Location          Neighbourhood  \
0          60 Skyview Ranch Drive NE                Skyview   
1           211- 355 Taralake Way NE               Taradale   
2         West Hillhurst Calgary, AB         West Hillhurst   
3           301 - 4603 Varsity Dr NW                Varsity   
4   20 & 40 Skyview Ranch Landing NE                Skyview   
5   20 & 40 Skyview Ranch Landing NE                Skyview   
6          60 Skyview Ranch Drive NE                Skyview   
7         216 Arbour Summit Place NW            Arbour Lake   
8          24 Mission Road Southwest  Parkhill-Stanley Park   
9                     1528 16 Ave NW           Capitol Hill   
10                     3704 20 St SW               Altadore   
11                    1922 17 ave sw               Scarboro   
12          104-19A Hemlock Cres. SW           Spruce Cliff   
13      L208, 1919 University Dr. NW     University Heights   
14      1010 First Street South West               Beltline   
15          2511 38 Street Northeast                 Rundle   
16               1536 45 street s.e,            Forest Lawn   
17                  937 37 Street SW            Rosscarrock   
18           Banff Trail Calgary, AB            Banff Trail   
19               7107 Elbow Drive SW           Kelvin Grove   

                                      Fulldescription  \
0   Join Us for Our Virtual Open House on Monday a...   
1   Large 2 bedroom condo with den ideal for small...   
2   Entire Cozy 2 Bedroom Cottage Bungalow in West...   
3   Luxurious condo living in 8 story concrete bui...   
4   Join Us for Our Virtual Open House on Monday a...   
5   Join Us for Our Virtual Open House on Monday a...   
6   Join Us for Our Virtual Open House on Monday a...   
7   6 Bedrooms, 3.5 Bathrooms, 1 Living Room, 1 Fa...   
8   Contemporary 2 bedroom / 2 bathroom corner uni...   
9   Take a virtual tour of ONE6 Residential and re...   
10  The apartment is within walking distance to Ma...   
11  Renting on approved credit only. Available imm...   
12  Rent today and save up to 2 months on rent wit...   
13  Rent today and save up to 2 months on rent wit...   
14  Lofts on First offers brand new, loft-style re...   
15  BEAUTIFUL 2 BEDROOM BUNGALOW TOWNHOUSE - STEPS...   
16  Renovated 4 yrs ago, new windows cut in for na...   
17  Westbrook Manor is professionally managed by M...   
18  Very beautiful & bright newly renovated 3 bedr...   
19  Mainstreet on Elbow is professionally managed ...   

                                       FlexTable_info  \
0   LISTING ID\n430647\nPETS\nCats & dogs negotiab...   
1   LISTING ID\n124512\nPETS\nNo pets\nSMOKING\nNo...   
2   LISTING ID\n437801\nPETS\nNo cats Dogs negotia...   
3   LISTING ID\n407306\nPETS\nNo pets\nSMOKING\nNo...   
4   LISTING ID\n430648\nPETS\nCats

In [4]:
# initial dimensions
df_rentfaster.shape

(1488, 7)

~1500 scraped entries are present in the raw data.

## Data Extraction from FlexTable_info

I defined a function for crude information extraction from Flex Table in the cell below.

In [5]:
def FlexTable_info_cleaning(Column, Keyword):
    """
    This function cleans the info in Flex Table. It takes
    a keyword and checks every row of pandas series for a match.
    The first word after the match and a space character is 
    stored in the column to be returned.
    """
    Cleaned_column = Column.apply(lambda x: x.partition(Keyword+'\n')[2].partition('\n')[0])
    return Cleaned_column

Storing the information extracted from Flex Table in new columns

In [6]:
# extracting the information from Flex Table and storing it in different columns
df['Listing_ID'] = FlexTable_info_cleaning(df['FlexTable_info'], 'LISTING ID')
df['Pets'] = FlexTable_info_cleaning(df['FlexTable_info'], 'PETS')
df['Smoking'] = FlexTable_info_cleaning(df['FlexTable_info'], 'SMOKING')
df['Tenant_Parking'] = FlexTable_info_cleaning(df['FlexTable_info'], 'TENANT PARKING')
df['Year_Built'] = FlexTable_info_cleaning(df['FlexTable_info'], 'YEAR BUILT')
df['Landlord_On_Rentfaster_Since'] = FlexTable_info_cleaning(df['FlexTable_info'], 'LANDLORD ON RENTFASTER SINCE')

## Data Extraction from FeaturesTable_info

Next, I defined a function for crude information extraction from Features Table.

In [7]:
def FeaturesTable_info_cleaning(Column, Keyword):
    """
    This function cleans the info in Features Table. It takes
    a keyword and checks every row of pandas series for a match.
    The first word after the match is 
    stored in the column to be returned.
    """
    Cleaned_column = Column.apply(lambda x: str(x).partition(Keyword)[2].partition('\n')[0])
    return Cleaned_column

Storing the information extracted from Features Table in new columns

In [8]:
# extracting the information from Features Table and storing it in different columns
df['Building_features'] = FeaturesTable_info_cleaning(df['FeaturesTable_info'], 'BUILDING FEATURES')
df['Property_features'] = FeaturesTable_info_cleaning(df['FeaturesTable_info'], 'PROPERTY FEATURES')
df['Community_features'] = FeaturesTable_info_cleaning(df['FeaturesTable_info'], 'COMMUNITY FEATURES')


In [9]:
# checking the dimensions of dataframe after adding columns 
df.shape

(1488, 16)

## Data Extraction from MainTable_info

Then I defined a function for crude information extraction from Main Info below.

In [10]:
def MainTable_info_cleaning(Column, Start_position):
    """
    This function cleans the info in Main Table. It takes
    a starting position and goes to every 6th position from
    there along the string obtained after splitting row at "'" 
    and append the value there to a list. For Main the starting 
    position is at 1 and for utilities info, it is at 5.
    """
    Cleaned_column = []
    for i in range(0, len(Column)):
        temp = []
        for j in range(Start_position, len(Column.apply(lambda x: str(x).split("'"))[i]), 6):
            temp.append((Column.apply(lambda x: str(x).split("'"))[i])[j])
        Cleaned_column.append(temp)
    return Cleaned_column

Storing the information extracted from Main Info in new columns

In [11]:
# extracting the information from Main Table and storing it in different columns
df['Main_info'] = MainTable_info_cleaning(df['MainTable_info'], 1)
df['Utilities_info'] = MainTable_info_cleaning(df['MainTable_info'], 5)

In [12]:
# checking the format of an entry
df['Main_info'][0]

['Apartment $1433 $1433 2 bd 1 ba 873 ft2 Unfurnished Long Term Immediate  ',
 'Apartment $1499 $1499 2 bd 2 ba 1015 ft2 Unfurnished Long Term Immediate  ',
 'Apartment $1348 $1348 1 bd 2 ba 946 ft2 Unfurnished Long Term Apr 01, 2021  ',
 'Apartment $1435 $1435 2 bd 1 ba 878 ft2 Unfurnished Long Term Apr 01, 2021  ']

This shows there are multiple sub-listings within a listing sometimes and thus, this column needs to be *exploded* (i.e. unpacked into different entries later) (<a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.explode.html" target="_blank">refer this link for more on exploding columns</a>). Refer the image below for similar listing example (from original website) with multiple sub-listings.

 ![explode_example](images/explode_example.png)

In [13]:
# checking the format of an entry
df['Utilities_info'][0]

['Unit Number: 100-310    Unit Description: Two Bedroom\\nUtilities Included: Heat Water',
 'Unit Number: 60-403    Unit Description: Two Bedroom\\nUtilities Included: Heat Water',
 'Unit Number: 100-222    Unit Description: One Bedroom\\nUtilities Included: Heat Water',
 'Unit Number: 60-410    Unit Description: Two Bedroom\\nUtilities Included: Heat Water']

In [14]:
# checking the dimensions of dataframe 
df.shape

(1488, 18)

I dropped the columns from which I extracted information in cell below.

In [15]:
# dropping the columns from which info extracted
df.drop(['FlexTable_info','FeaturesTable_info','MainTable_info'], axis=1, inplace=True)

In [16]:
# column number reduced
df.shape

(1488, 15)

## Exploding Data Frame Entries to Get Sub-Listings

As mentioned in previous section, each listing can have multiple sublistings. Therefore, it is necessary to explode the data frame along Main Info. This duplicates entries in every other column and for main info, it assigns every entry to a new row.
<a href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.explode.html" target="_blank">For more on exploding a column, check this link</a>.


In [17]:
# exploding the rows of dataframe along Main_info as every listing can have further sublistings
df_explode = df.explode('Main_info')
df_explode = df_explode.reset_index()
df_explode.rename(columns = {'index':'Old_Index'}, inplace = True)


In [18]:
# checking dimensions of exploded data frame
df_explode.shape

(4735, 16)

I reconfirmed above that the dataframe was exploded. Now we have sub-listings in different rows.

# Cleaning Data in Sub-Categories

After initial data cleaning in previous section, I extracted data from the sub-categories more rigorously.

## Cleaning Data in MainTable_info 

All the features in main info table were obtained one by one below.

### Cleaning Listing_Type Column

In [19]:
# reconfirming logic for a single element to get listing type
df_explode['Main_info'][1].partition(' ')[0]

'Apartment'

The first element after partitioning at empty space should be listing type by the logic above.

In [20]:
# first extracting information contained in Main_info table
# getting listing type 
df_explode['Listing_Type'] = df_explode['Main_info'].apply(lambda x: str(x).partition(' ')[0])

In [21]:
df_explode['Listing_Type'].unique()

array(['Apartment', 'Condo', 'House', 'Townhouse', 'Basement', 'Main',
       ',', 'Duplex', 'Shared', 'Loft', 'Parking', '', 'Neil',
       'Basement\\n$850', 'Office', 'Basement\\n$1150', 'Storage',
       'Basement\\n$1000', 's', 'Acreage'], dtype=object)

I checked the count for different listing types below. 

In [22]:
# a few of the listing types look off, otherwise mostly good  
df_explode['Listing_Type'].value_counts()

Apartment          3855
Condo               301
Townhouse           142
Basement            121
Shared               77
House                71
Main                 67
,                    44
Duplex               24
                     10
Office                6
Loft                  4
Storage               3
Neil                  2
Acreage               2
Parking               2
Basement\n$850        1
Basement\n$1000       1
s                     1
Basement\n$1150       1
Name: Listing_Type, dtype: int64

It is advisable to remove problematic listing type above before proceeding as there are only a few.  

This function gives the index of problematic listings at Nth position in value_counts() output above.

In [23]:
def Problematic_Index_Listing_Type(Column, N):
    """ This function returns index of problematic listing type at Nth position in value_counts()"""
    return (np.array(Column.value_counts().index))[N]

This function removes the problematic listings from dataframe.

In [24]:
def Remove_Problematic_Row(DataFrame, Column, Key):
    """ This function removes rows with problematic listing type as per value"""    
    return DataFrame.drop(DataFrame[(Column == Key)].index, inplace= True)
    

In the code block below, all the listing type having errors are cleaned up. 

In [25]:
# Getting index of problematic listing, trying new approach as compared to above
Problematic_1 = Problematic_Index_Listing_Type(df_explode['Listing_Type'], 9)
Problematic_2 = Problematic_Index_Listing_Type(df_explode['Listing_Type'], -1)
Problematic_3 = Problematic_Index_Listing_Type(df_explode['Listing_Type'], -3)
Problematic_4 = Problematic_Index_Listing_Type(df_explode['Listing_Type'], -4)

# Removing rows with problematic listing 
Remove_Problematic_Row(df_explode, df_explode.Listing_Type, Problematic_1)
Remove_Problematic_Row(df_explode, df_explode.Listing_Type, Problematic_2)
Remove_Problematic_Row(df_explode, df_explode.Listing_Type, Problematic_3)
Remove_Problematic_Row(df_explode, df_explode.Listing_Type, Problematic_4)

# Removing rows with problematic listing as per value
Remove_Problematic_Row(df_explode, df_explode.Listing_Type, ',')
Remove_Problematic_Row(df_explode, df_explode.Listing_Type, 's')


In [26]:
Problematic_1

''

The cell block above/below show that the code  is targeting the correct type of listings.

In [27]:
# printing to reconfirm
Problematic_3

'Basement\\n$1000'

In [28]:
# reindexing dataframe after using .drop()
df_explode = df_explode.reset_index(drop=True)

In [29]:
# listing type column cleaned here
df_explode['Listing_Type'].value_counts()

Apartment    3855
Condo         301
Townhouse     142
Basement      121
Shared         77
House          71
Main           67
Duplex         24
Office          6
Loft            4
Storage         3
Neil            2
Acreage         2
Parking         2
Name: Listing_Type, dtype: int64

Listing types look good to go.

### Cleaning Sq_Feet Column

It was noticed from trial and error that the value for size is located before the first ' ft' occurance and after the first empty space from reverse.

In [30]:
# after listing type, next getting information on space for every individual listing in sq feet
df_explode['Sq_Feet'] = df_explode['Main_info'].apply(lambda x: str(x).partition(' ft')[0].rpartition(' ')[2])

In [31]:
# Immediate, empty and Vacancy are obvious errors. The few remaining errors are dealt with next
df_explode['Sq_Feet'].value_counts()[:50]

Immediate    511
             248
608          158
909           90
936           80
593           61
564           61
914           61
935           60
800           51
Vacancy       48
600           48
900           48
727           45
550           41
923           40
636           40
589           39
850           39
1100          39
774           38
1000          38
650           34
748           34
426           33
563           32
700           32
545           32
836           30
657           28
710           28
896           28
685           27
913           26
725           26
1200          26
580           25
884           25
567           24
830           24
880           23
612           23
500           22
1030          22
885           21
635           21
587           20
1221          20
716           20
531           20
Name: Sq_Feet, dtype: int64

First, taking care of the issues with small value_counts() and removing problematic rows.

In [32]:
# intial dimension
print(df_explode.shape)

# Removing rows with problematic listing as per value
Remove_Problematic_Row(df_explode, df_explode.Sq_Feet, 'Water"]')
Remove_Problematic_Row(df_explode, df_explode.Sq_Feet, 'Availability')
Remove_Problematic_Row(df_explode, df_explode.Sq_Feet, 'sq.')
Remove_Problematic_Row(df_explode, df_explode.Sq_Feet, 'Sq.')
Remove_Problematic_Row(df_explode, df_explode.Sq_Feet, 'sq')

# reset index
df_explode = df_explode.reset_index(drop=True)

# 18 rows dropped in total
print(df_explode.shape)

(4677, 18)
(4659, 18)


18 rows dropped in total

In [33]:
df_explode['Sq_Feet'].value_counts()[:50]

Immediate    511
             248
608          158
909           90
936           80
564           61
593           61
914           61
935           60
800           51
900           48
Vacancy       48
600           48
727           45
550           41
923           40
636           40
1100          39
850           39
589           39
1000          38
774           38
748           34
650           34
426           33
700           32
545           32
563           32
836           30
710           28
657           28
896           28
685           27
913           26
725           26
1200          26
884           25
580           25
567           24
830           24
880           23
612           23
1030          22
500           22
635           21
885           21
1004          20
587           20
527           20
531           20
Name: Sq_Feet, dtype: int64

I then defined a function to change value at every row position, where there is a particular pattern, to a fixed value.

In [34]:
def Specific_Row_Value_Change(DataFrame, Column, Pattern, To_Be_Value):
    """
    This function changes initial value to To_Be_Value for all entries 
    in a data frame column of particular pattern in value_counts()
    """
    df_temp = np.array(Column)
    df_temp[DataFrame[(Column == Pattern)].index.to_list()] = To_Be_Value 
    Column = pd.Series(df_temp)
    
    return Column

These 3 type of errors (refer cell below) are occuring because people have not mentioned size of their listing (and last word in the thread is being picked up). Therefore, replacing -1 for size at these positions.

In [35]:
# removing problematic rows in value_count() above
df_explode.Sq_Feet = Specific_Row_Value_Change(df_explode, df_explode.Sq_Feet, 'Vacancy', '-1')
df_explode.Sq_Feet = Specific_Row_Value_Change(df_explode, df_explode.Sq_Feet, '', '-1')
df_explode.Sq_Feet = Specific_Row_Value_Change(df_explode, df_explode.Sq_Feet, 'Immediate', '-1')

In [36]:
df_explode['Sq_Feet'].value_counts()[:50]

-1      807
608     158
909      90
936      80
593      61
564      61
914      61
935      60
800      51
900      48
600      48
727      45
550      41
636      40
923      40
589      39
850      39
1100     39
774      38
1000     38
650      34
748      34
426      33
700      32
563      32
545      32
836      30
657      28
896      28
710      28
685      27
725      26
913      26
1200     26
884      25
580      25
830      24
567      24
880      23
612      23
500      22
1030     22
885      21
635      21
527      20
587      20
687      20
531      20
716      20
1004     20
Name: Sq_Feet, dtype: int64

This column looks good for now.

### Cleaning Furnishing Column 

Breaking at 'urn' pattern as it is unique and taking the sequence before that.

In [37]:
# next extracting info on whether furnished or not from Main_info. Breaking at 'urn' and taking value before that
df_explode['Furnishing'] = df_explode['Main_info'].apply(lambda x: str(x).rpartition('urn')[0].split(' ')[-1])
df_explode['Furnishing'].value_counts()

Unf    4291
F       256
        112
Name: Furnishing, dtype: int64

Unf stands for Unfurnished and was assigned 0. F stands for Furnished and was assigned 1. Rest are empty. Thus, taken care of accordingly (-1 inserted).

In [38]:
# assigned values 1 for furnished, 0 for unfurnished and -1 for not given
df_explode.Furnishing = Specific_Row_Value_Change(df_explode, df_explode.Furnishing, 'F', '1')
df_explode.Furnishing = Specific_Row_Value_Change(df_explode, df_explode.Furnishing, '', '-1')
df_explode.Furnishing = Specific_Row_Value_Change(df_explode, df_explode.Furnishing, 'Unf', '0')

In [39]:
# furnishing information clean
df_explode['Furnishing'].value_counts()

0     4291
1      256
-1     112
Name: Furnishing, dtype: int64

Information on furnishing is clean enough now.

### Cleaning Baths Column     

Breaking at first occurance of ba from behind and taking in everything till bd is reached. This is done as many listings have other words like den as well in between.

In [40]:
# extracting information regarding number of baths the Main_info
# splitting done in order to accomodate for den info which some of the listings had 
df_explode['Baths'] = df_explode['Main_info'].apply(lambda x: str(x).rpartition('ba')[0].partition('bd')[2])

In [41]:
# need to handle junk 
# also extract info from last for listings having 'den'
df_explode['Baths'].value_counts()

 1                                    2761
 2                                    1068
 + den 1                               241
 1.5                                   181
                                       168
 + den 2                               105
 2.5                                    65
 + den 2.5                              20
 3                                      12
 3.5                                    11
 + den 1.5                               8
 + den 3                                 6
 + den 3.5                               4
 + den 4.5                               2
 4                                       2
 4.5                                     2
 + den 1.5 ba 968 (main) + 750 (         1
 2 ba 1400 approximately plus 400        1
 6                                       1
Name: Baths, dtype: int64

Presently, not including the number of dens in analysis and therefore discarding that information below.

In [42]:
# extracting info from listings having den info
#replacing empty with -1
df_explode.Baths = Specific_Row_Value_Change(df_explode, df_explode.Baths, '', '-1')
df_explode.Baths = Specific_Row_Value_Change(df_explode, df_explode.Baths, ' + den 1 ', ' 1 ')
df_explode.Baths = Specific_Row_Value_Change(df_explode, df_explode.Baths, ' + den 2 ', ' 2 ')
df_explode.Baths = Specific_Row_Value_Change(df_explode, df_explode.Baths, ' + den 2.5 ', ' 2.5 ')
df_explode.Baths = Specific_Row_Value_Change(df_explode, df_explode.Baths, ' + den 1.5 ', ' 1.5 ')
df_explode.Baths = Specific_Row_Value_Change(df_explode, df_explode.Baths, ' + den 3 ', ' 3 ')
df_explode.Baths = Specific_Row_Value_Change(df_explode, df_explode.Baths, ' + den 3.5 ', ' 3.5 ')
df_explode.Baths = Specific_Row_Value_Change(df_explode, df_explode.Baths, ' + den 4.5 ', ' 4.5 ')

#not counting dens anywhere right now maybe if interested can do it sometime later


Now, checking the values in Baths column.

In [43]:
#  just a few junk values left to clean-up
df_explode['Baths'].value_counts()

 1                                    3002
 2                                    1173
 1.5                                   189
-1                                     168
 2.5                                    85
 3                                      18
 3.5                                    15
 4.5                                     4
 4                                       2
 6                                       1
 2 ba 1400 approximately plus 400        1
 + den 1.5 ba 968 (main) + 750 (         1
Name: Baths, dtype: int64

Dropping two rows for the problematic listing at the end of value_counts() 

In [44]:
# getting index for 2 problematic listing above
Problematic_5 = Problematic_Index_Listing_Type(df_explode['Baths'], -1)
Problematic_6 = Problematic_Index_Listing_Type(df_explode['Baths'], -2)

# removing problematic bath types from above
Remove_Problematic_Row(df_explode, df_explode.Baths, Problematic_5)
Remove_Problematic_Row(df_explode, df_explode.Baths, Problematic_6)


In [45]:
# reseting index as dropped rows 
df_explode = df_explode.reset_index(drop=True)

In [46]:
# column having info on number of baths is clean now
df_explode['Baths'].value_counts()

 1       3002
 2       1173
 1.5      189
-1        168
 2.5       85
 3         18
 3.5       15
 4.5        4
 4          2
 6          1
Name: Baths, dtype: int64

Number of baths is cleaned up.

### Cleaning Beds Column

Extracting information regarding number of beds

In [47]:
# extracting information regarding no. of beds from Main_info column
df_explode['Beds'] = df_explode['Main_info'].apply(lambda x: str(x).rpartition(
    'bd')[0].partition(' ')[2].partition(' ')[2].partition(' ')[2])

In [48]:
# mostly looks correct
df_explode['Beds'].value_counts()

1                                       1936
2                                       1860
3                                        274
                                         166
month 2                                  155
month 1                                   89
4                                         30
month 3                                   21
5                                         16
6                                          4
$1250 3                                    3
$950 2                                     3
$1300 3                                    3
Month Rent 2                               3
month rent 2                               3
as rent 2                                  3
$1500 3                                    3
month rent 1                               3
month rent 3                               2
OFF 2                                      2
1350.00 2                                  2
$1375 3                                    2
$1400 2   

Making the empty rows -1.

In [49]:
# making the empty rows -1
df_explode.Beds = Specific_Row_Value_Change(df_explode, df_explode.Beds, '', '-1')

In [50]:
# empty rows changed
df_explode['Beds'].value_counts()

1                                       1936
2                                       1860
3                                        274
-1                                       166
month 2                                  155
month 1                                   89
4                                         30
month 3                                   21
5                                         16
6                                          4
Month Rent 2                               3
as rent 2                                  3
$1250 3                                    3
month rent 2                               3
$950 2                                     3
$1300 3                                    3
$1500 3                                    3
month rent 1                               3
$1200 2                                    2
$1400 3                                    2
$2000 3                                    2
Months Rent 2                              2
1350.00 2 

Taking a very simple approach to cleaning. Considering only those value_counts() above for which counts is less than 156. For all these, taking the last character of the string as that contains required information.

In [51]:
# handling all the types with value count less than 156 in one go separately in temporary data frame 
# extracting the last digit for every row, which is basically no. of beds
counts = df_explode['Beds'].value_counts() 
df_temp = pd.Series(df_explode[df_explode['Beds'].isin(counts[counts < 156].index)]['Beds'])
# padding up to ensure matches with the format of the previous one
df_temp = (df_temp.str[-2]).str.pad(2, side='right', fillchar=' ')


In [52]:
# for debugging purposes
# len(df_temp)
# df_temp.unique()
# df_temp.value_counts()


In [53]:
# updating all rows where the value type belongs to a category 
# that has a count less than 155 in whole column of beds
df_explode['Beds'].update(df_temp)
df_explode['Beds'].value_counts()
# no of beds information cleaned

2     2059
1     2039
3      337
-1     166
4       34
5       17
6        5
Name: Beds, dtype: int64

Successfully parsed information on number of beds above.

### Cleaning Rent_Monthly Column

The function below gives the number of entries with N dollar signs in the string and their indices. Idea is to get a better understanding regarding number of dollar signs in listings in the dataset.

In [54]:
def Count_no_of_Dollars(DataFrame, N, Column):
    """
    This function return the no. of entries and the 
    index of all the entries with N number of dollar signs
    """
    Index_Dollar_Sign = DataFrame.index.where(pd.Series(list(map(lambda x: len(x.split('$')), Column))) == N).tolist()
    Quantity_Dollar_Sign = DataFrame.index.where(pd.Series(list(map(lambda x: len(x.split('$')), Column))) == N).notnull().sum()
    return Index_Dollar_Sign , Quantity_Dollar_Sign


In [55]:
Index_0_Dollar_Sign, Quantity_0_Dollar_Sign = Count_no_of_Dollars(df_explode, 1, df_explode['Main_info'])
Index_1_Dollar_Sign, Quantity_1_Dollar_Sign = Count_no_of_Dollars(df_explode, 2, df_explode['Main_info'])
Index_2_Dollar_Sign, Quantity_2_Dollar_Sign = Count_no_of_Dollars(df_explode, 3, df_explode['Main_info'])
Index_3_Dollar_Sign, Quantity_3_Dollar_Sign = Count_no_of_Dollars(df_explode, 4, df_explode['Main_info'])

# uncomment to check
# Quantity_0_Dollar_Sign
# df_explode.shape



I learnt that there are 0 entries with 0 dollar sign, 428 entries with 1 dollar sign, 4220 entries with 2 dollar signs, 9 entries with 3 dollar signs. All listings (4657) are fully accounted for till here.


Breaking at first occurance of dollar sign as it is for rent amount.

In [56]:
# extracting info regarding rent from Main_info table
df_explode['Rent_Monthly'] = df_explode['Main_info'].apply(lambda x: str(x).partition('$')[2].partition(' ')[0])

In [57]:
# rental info looks clean
df_explode['Rent_Monthly'].value_counts()[:50]

1425    219
1375    169
1250    160
1200    132
1100    127
1400    119
1225    109
1450     98
1350     92
1150     83
1300     80
1475     79
1275     76
950      69
1175     68
1050     59
1125     55
1500     53
1525     47
1600     45
1800     45
1750     41
925      36
1900     36
1550     35
1700     35
1650     34
1000     34
999      34
1625     32
900      31
1075     29
1025     29
1399     28
1850     27
1139     26
1195     26
2200     25
2150     24
2250     24
1245     24
1295     23
1229     23
850      22
800      22
975      22
2750     22
1172     22
1325     22
3150     21
Name: Rent_Monthly, dtype: int64

I removed problematic rows with 3 dollar signs. Rows with 1 dollar sign have only rent information and thus no modification was required for them. Rows with 2 dollar signs are inherently correct.

In [58]:
# removing 9 rows having 3 dollar signs
# row having 2 dollar signs by default correct
# rows having only 1 dollar sign are assumed to be having rent value but no deposit info 
df_explode.drop(pd.Series(Index_3_Dollar_Sign).value_counts().index.to_list(), inplace= True)
# reseting index
df_explode = df_explode.reset_index(drop=True)

In [59]:
# 9 rows dropped from data frame
df_explode.shape

(4648, 22)

Montlhy rent information is collected and cleaned here.

### Cleaning Utilities_info Column   

Taking the Utilities info after the keyword 'Included: '

In [60]:
# extracting utilities info from Utilities_info
df_explode['Utilities_info'] = df_explode['Utilities_info'].apply(
    lambda x: str(x).partition('Included: ')[2].partition("'")[0])

In [61]:
# types mostly make sense 
df_explode['Utilities_info'].value_counts()

Heat Water                               3157
                                          527
Water Heat                                135
Heat Water Electricity                    125
Heat Electricity Water                    102
Cable Heat Water                          100
Heat Cable Electricity Internet Water      62
Heat Cable Water                           61
Water                                      57
Heat                                       52
Heat Electricity Water Internet            44
Cable Heat Water Electricity               42
Heat Cable Internet Water                  31
Heat Electricity Internet Water            25
Heat Electricity Water Cable Internet      21
Water Heat",                                9
Internet                                    9
Heat Water Internet                         7
Heat Internet Electricity Water             5
Heat Water Electricity Internet             5
Heat Internet                               4
Heat Electricity Water Internet Ca

Changing empty to -1.

In [62]:
df_explode.Utilities_info = Specific_Row_Value_Change(df_explode, df_explode.Utilities_info, '', '-1')

In [63]:
# updated -1 at empty
df_explode['Utilities_info'].value_counts()

Heat Water                               3157
-1                                        527
Water Heat                                135
Heat Water Electricity                    125
Heat Electricity Water                    102
Cable Heat Water                          100
Heat Cable Electricity Internet Water      62
Heat Cable Water                           61
Water                                      57
Heat                                       52
Heat Electricity Water Internet            44
Cable Heat Water Electricity               42
Heat Cable Internet Water                  31
Heat Electricity Internet Water            25
Heat Electricity Water Cable Internet      21
Water Heat",                                9
Internet                                    9
Heat Water Internet                         7
Heat Internet Electricity Water             5
Heat Water Electricity Internet             5
Heat Electricity Water Internet Cable       4
Heat Internet                     

This function displays the words mentioned in value_counts() and their frequency. This helps in deciding which all words are frequently mentioned in the given field.

In [64]:
def Display_Split_ValueCountsIndex_Frequency (Column):
    
    """
    This function displays the frequency of elements 
    (generated by splitting string in value_counts() output at ' ') 
    in descending order. Gives us an idea of the most 
    frequent words in the value_counts().
    """

# converting to pandas series 
    temp_1 = pd.Series(Column.value_counts().index)
# initializing list for copying split elements
    temp_2 = []
    for x in range(0,len(temp_1)):
            temp_2.append(temp_1.str.split(' ')[x])
# ref: XXX
# got unique entries and their counts
    chain = itertools.chain(*temp_2)
    unique, counts = np.unique(list(chain), return_counts=True)
# storing results in a dictionary 
    dict = {}
    for A, B in zip(unique, counts):
        dict[A] = B
# returning dictionary sorted in reverse order as per frequency
    return sorted(dict.items(), key = lambda x: x[1], reverse= True)
    

Observing what is mentioned most in value_counts() for Utilities_info

In [65]:
# displaying most mentioned words in value_counts() with their frequency
Display_Split_ValueCountsIndex_Frequency(df_explode['Utilities_info'])

[('Heat', 57),
 ('Water', 56),
 ('Internet', 43),
 ('Electricity', 42),
 ('Cable', 34),
 ('', 3),
 ('Water",', 2),
 ('-1', 1),
 ('Heat",', 1)]

This function assigns score to every element depending on how many words from desired list it contains. +1 is assigned for every feature present.

In [66]:
def Specific_Row_Value_Change_Substring_PlusOne( Column, List_Substring):
    
    """
    This function assigns value to every listing.
    The listing having more of the desireable charactersistics 
    is rewarded more. For every match to a substring in List_Substring
    a +1 is added to the counter, which stores the final value.
    """
    temp_1 = {Substring:[] for Substring in List_Substring}
    for Substring in List_Substring:
        temp_2 = []
        for i in range(0, len(Column)):
            if (Column.str.contains(Substring).to_list()[i] == True):
                temp_2.append(1)
            else:   
                temp_2.append(0)
        temp_1[Substring] = temp_2        
    counter = np.zeros(len(Column))
    for Substring in List_Substring:
        counter = counter + np.array(temp_1[Substring])
                
    return pd.Series(counter)


Converting info in Utilities column to score based on how many desireable characteristics it has.

In [67]:
# checking the listing for how many desireable utility features it has
list_full_utilities_info = ['Heat', 'Water', 'Internet',
                            'Electricity', 'Cable', 'Water",', 'Heat",']
# giving a rating as per the no of matches, +1 for every match
df_explode['Utilities_info'] = Specific_Row_Value_Change_Substring_PlusOne(
    df_explode['Utilities_info'], list_full_utilities_info)

In [68]:
# utilities info looks cleaned up
df_explode['Utilities_info'].value_counts()

2.0    3304
0.0     527
3.0     417
4.0     170
1.0     119
5.0     111
Name: Utilities_info, dtype: int64

Making empty entries -1 again.

In [69]:
# replacing -1 at the 527 positions that were originally empty
df_explode.Utilities_info = Specific_Row_Value_Change(df_explode, df_explode.Utilities_info, 0.0, -1)
df_explode['Utilities_info'].value_counts()

 2.0    3304
-1.0     527
 3.0     417
 4.0     170
 1.0     119
 5.0     111
Name: Utilities_info, dtype: int64

## Cleaning Data in FlexTable_info 

Information from main info table was cleaned in the previous section. All the features from flex table were obtained one by one in the given section.

### Cleaning Neighbourhood Column

Neighbourhood column looks ok in principle. Issue is abnormally high number of listings in Applewood. Moreover, it is not even located close to downtown, which is the heart of the city, and thus there is not a clear explanation for so many listings in Appllewood. One of the possible explanations is that listings are repeated in the raw scraped data. Checking and correcting for repetition in Section 6.

In [70]:
# df_explode.shape
# df_explode.head()

In [71]:
# Applewood surprising, not even downtown where most of the listings should be
df_explode['Neighbourhood'].value_counts()

Applewood                987
Beltline                 351
Downtown                 330
Hillhurst                170
Lower Mount Royal        147
Mission                  141
Capitol Hill             131
Skyview                  119
Victoria Park            114
Kingsland                112
Tuxedo Park              103
Crescent Heights          97
Bankview                  94
Connaught                 90
South Calgary             85
Sunalta                   69
Varsity                   66
Richmond/Knob Hill        63
Eau Claire                53
Brentwood                 45
West Springs              44
Inglewood                 42
Radisson Heights          40
Kincora                   39
Haysboro                  38
Bridgeland                34
-1                        32
Spruce Cliff              31
Greenview                 31
Bowness                   29
Windsor Park              28
Bayview                   27
University Heights        27
Cornerstone               27
Cougar Ridge  

In [72]:
 len(df_explode['Neighbourhood'].unique())

167

There are 167 neighbourhoods in the city.

### Cleaning Smoking Column

In [73]:
# moving on to cleaning smoking status column
df_explode['Smoking'].value_counts()
# df_explode['Smoking'].unique()

Non-smoking            4146
Smoking allowed         307
Smoke free building     168
Smoking negotiable       15
                         12
Name: Smoking, dtype: int64

Assigning value 1 if allowed/negotiable, 0 if not allowed, -1 if not given.

In [74]:
#  if smoking allowed/negotiable then assign 1, else 0 and -1 for empty values
df_explode.Smoking = Specific_Row_Value_Change(df_explode, df_explode.Smoking, 'Non-smoking', '0')
df_explode.Smoking = Specific_Row_Value_Change(df_explode, df_explode.Smoking, 'Smoke free building', '0')
df_explode.Smoking = Specific_Row_Value_Change(df_explode, df_explode.Smoking, 'Smoking allowed', '1')
df_explode.Smoking = Specific_Row_Value_Change(df_explode, df_explode.Smoking, 'Smoking negotiable', '1')
df_explode.Smoking = Specific_Row_Value_Change(df_explode, df_explode.Smoking, '', '-1')

In [75]:
# smoking information cleaned up
df_explode['Smoking'].value_counts()

0     4314
1      322
-1      12
Name: Smoking, dtype: int64

Smoking information is clean now. -1 is for missing entries.

### Cleaning Landlord_On_Rentfaster_Since Column

In [76]:
# checking values in Landlord_On_Rentfaster_Since column
df_explode['Landlord_On_Rentfaster_Since'].value_counts()
# df_explode['Landlord_On_Rentfaster_Since'].unique()


        2192
2017    1049
2020     342
2019     232
2009     170
2018     143
2006      80
2015      79
2016      63
2014      59
2005      58
2008      35
2007      35
2011      31
2010      31
2013      28
2012      21
Name: Landlord_On_Rentfaster_Since, dtype: int64

Replacing empty with -1.

In [77]:
df_explode.Landlord_On_Rentfaster_Since = Specific_Row_Value_Change(
    df_explode, df_explode.Landlord_On_Rentfaster_Since, '', '-1')

In [78]:
# column looks clean
df_explode['Landlord_On_Rentfaster_Since'].value_counts()

-1      2192
2017    1049
2020     342
2019     232
2009     170
2018     143
2006      80
2015      79
2016      63
2014      59
2005      58
2007      35
2008      35
2010      31
2011      31
2013      28
2012      21
Name: Landlord_On_Rentfaster_Since, dtype: int64

### Cleaning Year_Built Column

In [79]:
# checking year built column next
df_explode['Year_Built'].value_counts()
# interestingly one building built in future XD

                                                             1998
2016                                                          741
2015                                                          421
2019                                                          300
2020                                                          283
2017                                                          264
2021                                                          132
1991                                                           55
2018                                                           45
2002                                                           25
2008                                                           21
2010                                                           20
2007                                                           16
2005                                                           15
1972                                                           14
1965      

In [80]:
# replacing empty with -1
df_explode.Year_Built = Specific_Row_Value_Change(df_explode, df_explode.Year_Built, '', '-1')


In [81]:
# replaced with -1
df_explode['Year_Built'].value_counts()

-1                                                           1998
2016                                                          741
2015                                                          421
2019                                                          300
2020                                                          283
2017                                                          264
2021                                                          132
1991                                                           55
2018                                                           45
2002                                                           25
2008                                                           21
2010                                                           20
2007                                                           16
2005                                                           15
1972                                                           14
1965      

All the problematic listings have value_count < 1. Handling them separately here. Making them all -1 (i.e. assuming not given). This is reasonable since there are only 28 defective listing out of ~4.5k total.

In [82]:
# can replace the ones less than 1 value count by -1 

counts = df_explode['Year_Built'].value_counts() 
df_temp = pd.Series(df_explode[df_explode['Year_Built'].isin(counts[counts < 2].index)]['Year_Built'])

y = df_temp.index.to_list()
for i in range(0, len(y)):
    df_explode['Year_Built'][y[i]] = '-1'




In [83]:
# year built cleaned
df_explode['Year_Built'].value_counts()

-1      2026
2016     741
2015     421
2019     300
2020     283
2017     264
2021     132
1991      55
2018      45
2002      25
2008      21
2010      20
2007      16
2005      15
1972      14
1965      13
1980      12
1985      12
2014      11
2013      11
1978      11
2003      10
2001      10
2004       9
2006       8
2011       8
1970       8
1962       8
2012       8
2000       7
1910       7
2009       7
1998       6
1976       6
1975       6
1996       5
1973       4
1969       4
1999       4
1979       4
1912       4
1981       4
1960       3
1977       3
1971       3
1988       3
1992       3
1950       3
1995       3
1911       3
1974       3
1967       3
1983       3
1997       3
1954       3
1990       3
1982       3
1949       2
1958       2
1964       2
1953       2
1947       2
1963       2
1984       2
1968       2
1957       2
Name: Year_Built, dtype: int64

### Cleaning Pets Column

There are 106 types of different listing so I needed a simplistic approach to clean it up.

In [84]:
# checking info in column related to pets
df_explode['Pets'].value_counts()

No pets                                                                  1469
Cats & dogs negotiable                                                   1451
Cats & dogs negotiable Fee: INCLUDED!!                                    361
Cats & dogs negotiable Fee: $35 per Month                                 285
Cats negotiable No dogs                                                   120
Cats & dogs negotiable Fee: 25                                            101
Cats & dogs negotiable Fee: $150/cat | $250/dog                            90
Cats & dogs negotiable Fee: Monthly Fee from $35                           86
Cats & dogs negotiable Fee: $30.00 per month /per pet                      81
Cats & dogs allowed                                                        78
Cats & dogs negotiable Fee: $300                                           65
Cats negotiable No dogs Fee: $25/month fee                                 55
Cats & dogs negotiable Fee: $0.00                               

In [85]:
df_explode.Pets = Specific_Row_Value_Change(df_explode, df_explode.Pets, '', 'No')

In [86]:
df_explode['Pets'].value_counts()

No pets                                                                  1469
Cats & dogs negotiable                                                   1451
Cats & dogs negotiable Fee: INCLUDED!!                                    361
Cats & dogs negotiable Fee: $35 per Month                                 285
Cats negotiable No dogs                                                   120
Cats & dogs negotiable Fee: 25                                            101
Cats & dogs negotiable Fee: $150/cat | $250/dog                            90
Cats & dogs negotiable Fee: Monthly Fee from $35                           86
Cats & dogs negotiable Fee: $30.00 per month /per pet                      81
Cats & dogs allowed                                                        78
Cats & dogs negotiable Fee: $300                                           65
Cats negotiable No dogs Fee: $25/month fee                                 55
Cats & dogs negotiable Fee: $0.00                               

Doing a very simplistic cleaning here. If the listing mentions No anywhere then assigned 0 (since I want to keep both cats and dogs or none). For all others, 1 assigned. Monetary information included in some of the listings above is not included in analysis to simplify process.

In [87]:
# any entry with No in field given 0, rest all given 1 
df_explode['Pets'][~df_explode['Pets'].str.contains('No')] = '1'
df_explode['Pets'][df_explode['Pets'].str.contains('No')] = '0'

In [88]:
# pets column cleaned up
df_explode['Pets'].value_counts()

1    2899
0    1749
Name: Pets, dtype: int64

1 signifies listings that allow pets and 0 signifies pets not allowed.

### Cleaning Tenant_Parking Column

In [89]:
# getting overview of different kind of entries mentioned in Tenant_Parking column
df_explode['Tenant_Parking'].value_counts()

Underground Fee: 1 STALL INCLUDED!!                                                          761
Outdoor                                                                                      578
Underground                                                                                  555
Street                                                                                       426
Covered                                                                                      233
Underground Street Fee: 1 STALL INCLUDED!!                                                   225
Underground Fee: $100                                                                        121
Outdoor Outdoor                                                                              115
Single garage Fee: 50/85/125                                                                 100
Underground Fee: $150/month                                                                   90
Underground Outdoor           

Displaying most mentioned words in the value_counts() for tenant parking in cell below.

In [90]:
# displaying most mentioned words in value_counts() with their frequency
Display_Split_ValueCountsIndex_Frequency(df_explode['Tenant_Parking'])

[('Fee:', 107),
 ('Outdoor', 85),
 ('Underground', 70),
 ('Covered', 40),
 ('Street', 29),
 ('garage', 21),
 ('Single', 11),
 ('parking', 8),
 ('Double', 7),
 ('Included', 7),
 ('$100', 6),
 ('Free', 6),
 ('None', 5),
 ('included', 5),
 ('rent', 5),
 ('$0', 4),
 ('1', 4),
 ('75', 4),
 ('for', 4),
 ('per', 4),
 ('$25', 3),
 ('&', 3),
 ('150', 3),
 ('50', 3),
 ('FREE', 3),
 ('Fee', 3),
 ('IN', 3),
 ('Negotiable', 3),
 ('No', 3),
 ('Parking', 3),
 ('in', 3),
 ('month', 3),
 ('$150', 2),
 ('$75', 2),
 ('$75.00', 2),
 ('+', 2),
 ('100', 2),
 ('2', 2),
 ('50.00', 2),
 ('Additional', 2),
 ('BE', 2),
 ('BUILT', 2),
 ('GARAGE', 2),
 ('INCLUDED!!', 2),
 ('SPRING', 2),
 ('STALL', 2),
 ('WILL', 2),
 ('spot', 2),
 ('the', 2),
 ('underground', 2),
 ('', 1),
 ('$0.00', 1),
 ('$115.00', 1),
 ('$150/month', 1),
 ('$175', 1),
 ('$26', 1),
 ('$35-$65', 1),
 ('$45', 1),
 ('$50', 1),
 ('$50.00', 1),
 ('$65', 1),
 ('$85/month', 1),
 ('$89/Month', 1),
 ('$89/month', 1),
 ('(INCLUDED', 1),
 ('(full', 1),
 ('-

This function changes value in a binary format at rows which match any element in a list.

In [91]:
def Specific_Row_Value_Change_Substring( Column, List_Substring, To_Be_Value):
    
    """
    This function changes value of every element in the column.
    The entries where a word from List_Substring is found, To_Be_Value is assigned there.
    At all other positions, 0 is assigned.  
    """
    
    for Substring in List_Substring:
        for i in range(0, len(Column)):
# checking if substring is present             
            if (Column.str.contains(Substring).to_list()[i] == True):
                Column[i] = To_Be_Value
# assigned 0 to entries which do not contain any word in List_Substring                
    for i in range(0, len(Column)):
            if (Column[i] != To_Be_Value):
                Column[i] = 0

    return Column

Checking if the parking mentions any favourable feature (i.e. protected parking in form of underground, garage or covered). If yes, then assigned 1 else 0.

In [92]:
# checking if the listing has any desireable characteristics regarding parking options
list_full = ['Underground', 'Covered', 'garage', 'GARAGE', 'underground']
# giving 0/1 rating to every listing
df_explode['Tenant_Parking'] = Specific_Row_Value_Change_Substring( df_explode['Tenant_Parking'], list_full, 1)

df_explode['Tenant_Parking'].value_counts()
# XXXX explain rationale what we are doing here

1    3372
0    1276
Name: Tenant_Parking, dtype: int64

## Cleaning Data in FeaturesTable_info 

Information from flex info table was cleaned in the previous section. All the data from features table was obtained one by one in the given section. I defined the function *Specific_Row_Value_Change_Substring_Advanced* to score building, property and community features. If the listing had a match for ordinary feature (in *List_Substring_Give1*) then +1 was awarded, if the listing had a match for better feature (in *List_Substring_Give2*) then +2 was awarded and if the listing had a match for best feature (in *List_Substring_Give3*) then +3 was awarded. 

In [93]:
def Specific_Row_Value_Change_Substring_Advanced( Column, List_Substring_Give1, List_Substring_Give2, List_Substring_Give3):
    
    """
    This function assigns score to every listing.
    The listing having more of the desireable charactersistics 
    ares rewarded more. 
    
    """

    def HelperFunction(Column, List_Substring, Points):
        
        temp_1 = {Substring:[] for Substring in List_Substring}
        for Substring in List_Substring:
            temp_2 = []
            for i in range(0, len(Column)):
                if (Column.str.contains(Substring).to_list()[i] == True):
                    temp_2.append(Points)
                else:   
                    temp_2.append(0)
            temp_1[Substring] = temp_2        
        counter = np.zeros(len(Column))
        for Substring in List_Substring:
            counter = counter + np.array(temp_1[Substring])

        return counter
        
    counter_main = np.zeros(len(Column))
    counter_1 = HelperFunction(Column, List_Substring_Give1, 1) 
    counter_2 = HelperFunction(Column, List_Substring_Give2, 2)
    counter_3 = HelperFunction(Column, List_Substring_Give3, 3)
    counter_main = counter_1 + counter_2 + counter_3    
        
    return pd.Series(counter_main)

### Cleaning Building_features Column

In [94]:
# df_explode['Building_features'].value_counts()

Checking for most mentioned words in value_counts()

In [95]:
# displaying most mentioned words in value_counts() with their frequency
Display_Split_ValueCountsIndex_Frequency(df_explode['Building_features'])

[('', 288),
 ('Entry', 229),
 ('Secure', 229),
 ('Elevator', 201),
 ('Guest', 170),
 ('Management', 157),
 ('On-Site', 157),
 ('Parking', 141),
 ('Area', 132),
 ('Fitness', 132),
 ('Bike', 129),
 ('room', 129),
 ('Building', 128),
 ('Concrete', 128),
 ('Room', 111),
 ('Courtyard', 94),
 ('Party/Games', 92),
 ('Common', 72),
 ('Terrace', 72),
 ('Walk-Up', 67),
 ('Concierge', 51),
 ('Suite', 29),
 ('Sauna', 27),
 ('Pool', 26),
 ('Swimming', 26),
 ('Hot-Tub', 21),
 ('Steam', 19),
 ('Theatre', 14),
 ('views', 2),
 ('Access', 1),
 ('Array', 1),
 ('City', 1),
 ('Mountain', 1),
 ('Wheelchair', 1)]

Scoring every listing on the basis of how many desireable characteristics it has.

In [96]:
# checking the listing for how many desireable characteristics regarding building features it has

list_building_features_point1 = ['Secure', 'Elevator']
list_building_features_point2 = ['On-Site', 'Fitness', 'Party/Games', 'Terrace']
list_building_features_point3 = ['Concierge', 'Sauna', 'Swimming', 'Hot-Tub', 'Theatre', 'Steam']



# giving a score as per the no of matches, +1/+2/+3 depending on type of match
df_explode['Building_features'] = Specific_Row_Value_Change_Substring_Advanced(
    df_explode['Building_features'], list_building_features_point1, list_building_features_point2, list_building_features_point3)

In [97]:
df_explode['Building_features'].value_counts()
# many of the building features look empty but that is coz many are not apartment, so no building features for them


8.0     1072
0.0      967
2.0      841
5.0      377
9.0      238
13.0     200
3.0      184
6.0      177
4.0      164
1.0      161
10.0      94
16.0      49
18.0      36
7.0       22
22.0      21
11.0      15
19.0      11
14.0       5
12.0       4
25.0       3
15.0       2
17.0       2
20.0       1
28.0       1
21.0       1
Name: Building_features, dtype: int64

Building features score looks clean.

### Cleaning Community_features Column

In [98]:
# df_explode['Community_features'].value_counts()

Checking for most mentioned words in value_counts()

In [99]:
# displaying most mentioned words in value_counts() with their frequency
Display_Split_ValueCountsIndex_Frequency(df_explode['Community_features'])

[('', 91),
 ('Bus', 90),
 ('Center', 89),
 ('Playground/Park', 89),
 ('Shopping', 89),
 ('Bike', 87),
 ('Paths', 87),
 ('Complex', 66),
 ('Sports', 66),
 ('Pool', 63),
 ('Courts', 57),
 ('Tennis', 57),
 ('Library', 50),
 ('Public', 50),
 ('Access', 48),
 ('Train', 48),
 ('Course', 47),
 ('Golf', 47),
 ('River', 36),
 ('Lake', 26),
 ('Outdoor', 20)]

Scoring every listing on the basis of how many desireable characteristics it has.  

In [100]:
list_community_features_point1 = ['Bus', 'Shopping', 'Bike']
list_community_features_point2 = ['Sports', 'Pool', 'Tennis', 'Library', 'Train', 'Playground/Park']
list_community_features_point3 = ['Golf', 'River', 'Lake' ]



# giving a score as per the no of matches, +1/+2/+3 depending on type of match
df_explode['Community_features'] = Specific_Row_Value_Change_Substring_Advanced(
    df_explode['Community_features'], list_community_features_point1, list_community_features_point2, list_community_features_point3)

In [101]:
# community features look clean
df_explode['Community_features'].value_counts()

18.0    1776
9.0     1139
21.0     509
14.0     272
16.0     159
19.0     156
24.0     119
15.0      98
11.0      82
12.0      65
13.0      58
0.0       42
8.0       40
7.0       35
3.0       27
5.0       24
17.0      20
22.0      19
10.0       5
6.0        1
4.0        1
20.0       1
Name: Community_features, dtype: int64

All listings seem to be assigned community features score correctly.

### Cleaning Property_features Column

In [102]:
# df_explode['Property_features'].value_counts()

Displaying the most frequent words in value_counts()

In [103]:
# displaying most mentioned words in value_counts() with their frequency
Display_Split_ValueCountsIndex_Frequency(df_explode['Property_features'])

[('Floors', 1008),
 ('Laundry', 984),
 ('', 978),
 ('-', 968),
 ('Fridge', 938),
 ('Oven/Stove', 896),
 ('Dishwasher', 657),
 ('In', 639),
 ('Suite', 639),
 ('Microwave', 563),
 ('Coverings', 558),
 ('Window', 558),
 ('Flooring', 545),
 ('Balcony', 510),
 ('Storage', 389),
 ('Carpeted', 380),
 ('Laminate', 365),
 ('Tile', 358),
 ('Patio/Deck', 287),
 ('Elevator', 276),
 ('Hardwood', 263),
 ('views', 245),
 ('Entry', 235),
 ('Private', 235),
 ('Shared', 229),
 ('Backyard', 207),
 ('Fenced', 207),
 ('In-suite', 199),
 ('Fireplace', 196),
 ('Lockers', 190),
 ('Air', 187),
 ('Conditioning', 187),
 ('City', 186),
 ('Access', 167),
 ('Internet', 151),
 ('Disposal', 107),
 ('Coin/Card', 99),
 ('Linoleum', 97),
 ('Luxury', 90),
 ('Plank', 90),
 ('Vinyl', 90),
 ('Alarm', 87),
 ('Basement', 59),
 ('Mountain', 59),
 ('Walk-Out', 59),
 ('Unit', 50),
 ('Central', 45),
 ('Vac', 45),
 ('Entrance', 44),
 ('Zero-Step', 44),
 ('Fire-Pit', 43),
 ('Corner', 40),
 ('River', 37),
 ('Views', 37),
 ('Jetted',

Scoring every listing on the basis of how many desireable characteristics it has.

In [104]:
list_property_features_point1 = ['Laundry', 'Fridge', 'Oven/Stove', 'Dishwasher', 'Microwave', 'Vinyl', 'Lockers']
list_property_features_point2 = ['Alarm', 'Carpeted', 'Hardwood', 'Balcony', 'Storage', 'Fire-Pit', 'Linoleum']
list_property_features_point3 = ['Private', 'Backyard', 'Fireplace', 'Tub/Jacuzzi', 'Hot-Tub', 'Patio/Deck', 'Penthouse', 'Corner'  ]



# giving a score as per the no of matches, +1/+2/+3 depending on type of match
df_explode['Property_features'] = Specific_Row_Value_Change_Substring_Advanced(
    df_explode['Property_features'], list_property_features_point1, list_property_features_point2, list_property_features_point3)

In [105]:
# property features look clean
df_explode['Property_features'].value_counts()

12.0    989
10.0    651
7.0     310
9.0     281
3.0     267
6.0     253
13.0    235
5.0     224
16.0    187
11.0    167
14.0    159
8.0     153
4.0     145
20.0     86
17.0     80
25.0     63
1.0      62
15.0     60
26.0     58
21.0     56
0.0      52
23.0     37
2.0      24
18.0     21
19.0     16
22.0      7
32.0      1
24.0      1
31.0      1
28.0      1
35.0      1
Name: Property_features, dtype: int64

Property features score is assigned correctly.

# Removing Repeated Listings

While cleaning the data I noticed that there are abnormally high number of listings for Applewood neighbourhood and this suggested that there were repeated entries in the raw data. I identified and cleaned the repeated entries in this section.

## Understanding the Issue

In the example below, repeated entries can be seen.

In [106]:
# we can see many listings repeating here are in applewood
# uncomment to see
# df_explode.iloc[3188:3500]

The issue is that people are reposting their listings and overtime, they are accumulating at the higher numbered pages. The solution to remove these repeated postings is to first get the listing ids that are getting repeated. After that, only the first group of exploded sub-listings should be taken for a given listing id. All subsequent occurances of the given listing id should then be deleted. 

In [107]:
# listing id count in exploded dataframe
df_explode['Listing_ID'].value_counts()[:50]

329389    400
327366    361
327410    225
307606    121
433243    100
147339    100
444140     90
361795     72
417008     64
355606     60
368446     60
5766       55
410845     55
439983     49
31264      48
412577     48
360545     48
434562     42
393079     42
407788     40
443323     40
393080     36
290307     35
407156     30
416372     27
340111     27
111726     25
417059     25
422955     25
414193     25
391053     25
426571     25
365511     25
19655      24
377368     24
305440     22
29347      21
300688     21
343051     21
316344     20
438172     20
297054     20
374053     20
351340     20
422562     20
406865     18
342591     16
342680     16
309853     16
2245       16
Name: Listing_ID, dtype: int64

In [108]:
# listing id count in original dataframe
df['Listing_ID'].value_counts()[:50]
#  therefore can find no. of sublistings per listing, will be helpful in cleaning up repeated ones


329389    20
327366    19
327410    15
307606    11
147339    10
444140    10
433243    10
361795     8
417008     8
439983     7
393080     6
393079     6
434562     6
443323     5
5766       5
417059     5
111726     5
368446     5
351340     5
426571     5
414193     5
391053     5
365511     5
410845     5
290307     5
407156     5
355606     5
422955     5
407788     4
343194     4
19655      4
433935     4
309853     4
342680     4
2245       4
360461     4
329452     4
424797     4
424798     4
342591     4
316344     4
31264      4
297054     4
75289      4
422562     4
374053     4
377368     4
438172     4
412577     4
131523     4
Name: Listing_ID, dtype: int64

For example from the above two value counts it can be seen that there are 20 listings for 329389 and each has 20 sublistings and thus after exploding, there are 400 (20 X 20) such listing ids. The same logic can then be extended further for other listing ids. Thus, for proper data cleaning, first occurance of 20 sublisting should be kept and rest 380 should be discarded.

## Getting Number of Sub-Listings per Listing

The code block below stores the number of sub-listings per listing id in dictionary dict_3.

In [109]:
# concerned only about repeated entries
counts_1 = df['Listing_ID'].value_counts() 
counts_1 = counts_1[counts_1 > 1]

# concerned only about repeated entries
counts_2 = df_explode['Listing_ID'].value_counts() 
counts_2 = counts_2[counts_2 > 1]

# initializing dictionary 1/2 for storing listing ids and their counts
dict_1 = {}

for i in range(0, len(counts_1)):
    dict_1[counts_1.index[i]] = counts_1[i]

dict_2 = {}

for i in range(0, len(counts_2)):
    dict_2[counts_2.index[i]] = counts_2[i]    

#  dictionary 3 containes no of sublistings per listing id
dict_3 = {}
for item in dict_1.items():
    try:
        dict_3[item[0]] = (dict_2[item[0]]/item[1])
    except:
        continue
        


In [110]:
# reviewing to see if makes sense
# dict_1
# dict_2
dict_3

{'329389': 20.0,
 '327366': 19.0,
 '327410': 15.0,
 '307606': 11.0,
 '147339': 10.0,
 '444140': 9.0,
 '433243': 10.0,
 '361795': 9.0,
 '417008': 8.0,
 '439983': 7.0,
 '393080': 6.0,
 '393079': 7.0,
 '434562': 7.0,
 '443323': 8.0,
 '5766': 11.0,
 '417059': 5.0,
 '111726': 5.0,
 '368446': 12.0,
 '351340': 4.0,
 '426571': 5.0,
 '414193': 5.0,
 '391053': 5.0,
 '365511': 5.0,
 '410845': 11.0,
 '290307': 7.0,
 '407156': 6.0,
 '355606': 12.0,
 '422955': 5.0,
 '407788': 10.0,
 '343194': 4.0,
 '19655': 6.0,
 '433935': 4.0,
 '309853': 4.0,
 '342680': 4.0,
 '2245': 4.0,
 '360461': 4.0,
 '329452': 3.0,
 '424797': 4.0,
 '424798': 4.0,
 '342591': 4.0,
 '316344': 5.0,
 '31264': 12.0,
 '297054': 5.0,
 '75289': 4.0,
 '422562': 5.0,
 '374053': 5.0,
 '377368': 6.0,
 '438172': 5.0,
 '412577': 12.0,
 '131523': 4.0,
 '343049': 4.0,
 '383003': 3.0,
 '355717': 3.0,
 '146362': 3.0,
 '439462': 3.0,
 '2244': 3.0,
 '402020': 3.0,
 '2594': 3.0,
 '338360': 5.0,
 '435439': 4.0,
 '100159': 4.0,
 '312479': 3.0,
 '4001

Therefore, I first identified number of sub-listing per id.

## Getting Index at Repetitions and Cleaning

Uptill here, the listing ids of concern and the number of sub-listing per id is know. Using this information, the code block below generates a list of all the problematic (repeated) index of data frame _df_explode_ which are then dropped.

In [111]:
# this code block identifies the repetitions in the dataframe df_explode 

# initializing empty list
list_removal = []
# checking for all listing ids in dict_3
for item in dict_3.items():
# no of repetition (i.e. no of sublistings)
    rep = dict_3[item[0]]
    index = item[0]
# counter to break loop as soon as first time listing id is found     
    count = 0
    for j in range(0, len(df_explode['Listing_ID'])):
        if(count>0):
            break
        else:
            if(df_explode['Listing_ID'].iloc[j] == item[0]):
                count = 1
                for i in range(j+int(rep)+1, len(df_explode['Listing_ID'])):
# noting all repetions after first group of sublistings (from same original unexploded listing) into removal list                    
                    if(df_explode['Listing_ID'][i] == item[0]):
                        list_removal.append(i)
                    else:
                        continue
            else:
                continue


In [112]:
# no of repeated entries
len(list_removal)

2930

2930 entries were repeated and thus were removed.

In [113]:
# dropping all repeated rows
df_explode_norepeat = df_explode.drop(list_removal, axis = 0)

In [114]:
# reseting index
df_explode_norepeat = df_explode_norepeat.reset_index(drop=True)

Checking information in neighbourhood column now.

In [115]:
# neighbourhood clean and make sense now
df_explode_norepeat['Neighbourhood'].value_counts()

Beltline                 148
Downtown                 104
Lower Mount Royal         78
Applewood                 56
Bankview                  55
Mission                   50
Victoria Park             47
Skyview                   38
Crescent Heights          37
Hillhurst                 37
Sunalta                   37
Eau Claire                36
Kingsland                 33
South Calgary             30
Varsity                   29
Brentwood                 25
Radisson Heights          22
Bridgeland                21
Haysboro                  20
Capitol Hill              19
Bowness                   19
Connaught                 18
Sunnyside                 17
Spruce Cliff              17
Inglewood                 16
Rosscarrock               15
Richmond/Knob Hill        15
East Village              15
Cornerstone               15
Kincora                   15
Windsor Park              14
West Springs              14
University Heights        14
Banff Trail               14
Tuxedo Park   

The neighbourhood data now makes sense. The zones with higher density of listings are located near downtown, which is the heart of all the activity in city.

# Saving Final Dataframe

Checking dataframe now.

In [116]:
# dataframe so far
df_explode_norepeat[0:20]

Old_Index                                     Listings_Title  \
0           0                                      VISTA PHASE 1   
1           0                                      VISTA PHASE 1   
2           0                                      VISTA PHASE 1   
3           0                                      VISTA PHASE 1   
4           1  LARGE 2 BEDROOMS WITH DEN CONDO AT TARALAKE WA...   
5           2  COZY 2 BEDROOM COTTAGE HOUSE WITH HUGE FENCED ...   
6           3                          VARSITY TOWERS NORTH WEST   
7           4                                      VISTA PHASE 2   
8           4                                      VISTA PHASE 2   
9           4                                      VISTA PHASE 2   
10          4                                      VISTA PHASE 2   
11          4                                      VISTA PHASE 2   
12          5                                      VISTA PHASE 2   
13          7             6 BED,LIVING ROOM,FAMILY ROOM,3.5 BATH   
14          8        2 BED/2 BATH CONTEMPORARY CONDO IN PARKHILL   
15          9  MOVE IN NOW AND DON'T PAY UNTIL MAY 1 + GET AP...   
16          9  MOVE IN NOW AND DON'T PAY UNTIL MAY 1 + GET AP...   
17          9  MOVE IN NOW AND DON'T PAY UNTIL MAY 1 + GET AP...   
18          9  MOVE IN NOW AND DON'T PAY UNTIL MAY 1 + GET AP...   
19          9  MOVE IN NOW AND DON'T PAY UNTIL MAY 1 + GET AP...   

                            Location          Neighbourhood  \
0          60 Skyview Ranch Drive NE                Skyview   
1          60 Skyview Ranch Drive NE                Skyview   
2          60 Skyview Ranch Drive NE                Skyview   
3          60 Skyview Ranch Drive NE                Skyview   
4           211- 355 Taralake Way NE               Taradale   
5         West Hillhurst Calgary, AB         West Hillhurst   
6           301 - 4603 Varsity Dr NW                Varsity   
7   20 & 40 Skyview Ranch Landing NE                Skyview   
8   20 & 40 Skyview Ranch Landing NE                Skyview   
9   20 & 40 Skyview Ranch Landing NE                Skyview   
10  20 & 40 Skyview Ranch Landing NE                Skyview   
11  20 & 40 Skyview Ranch Landing NE                Skyview   
12  20 & 40 Skyview Ranch Landing NE                Skyview   
13        216 Arbour Summit Place NW            Arbour Lake   
14         24 Mission Road Southwest  Parkhill-Stanley Park   
15                    1528 16 Ave NW           Capitol Hill   
16                    1528 16 Ave NW           Capitol Hill   
17                    1528 16 Ave NW           Capitol Hill   
18                    1528 16 Ave NW           Capitol Hill   
19                    1528 16 Ave NW           Capitol Hill   

                                      Fulldescription Listing_ID Pets Smoking  \
0   Join Us for Our Virtual Open House on Monday a...     430647    1       0   
1   Join Us for Our Virtual Open House on Monday a...     430647    1       0   
2   Join Us for Our Virtual Open House on Monday a...     430647    1       0   
3   Join Us for Our Virtual Open House on Monday a...     430647    1       0   
4   Large 2 bedroom condo with den ideal for small...     124512    0       0   
5   Entire Cozy 2 Bedroom Cottage Bungalow in West...     437801    0       0   
6   Luxurious condo living in 8 story concrete bui...     407306    0       0   
7   Join Us for Our Virtual Open House on Monday a...     430648    1       0   
8   Join Us for Our Virtual Open House on Monday a...     430648    1       0   
9   Join Us for Our Virtual Open House on Monday a...     430648    1       0   
10  Join Us for Our Virtual Open House on Monday a...     430648    1       0   
11  Join Us for Our Virtual Open House on Monday a...     430648    1       0   
12  Join Us for Our Virtual Open House on Monday a...     430648    1       0   
13  6 Bedrooms, 3.5 Bathrooms, 1 Living Room, 1 Fa...     444203    1       0   
14  Contemporary 2 bedroom / 2 bathr

The obtained dataframe looks clean and in the desired format. Dropping columns which would be not required in further analysis down below.

In [117]:
df_explode_norepeat.drop(['Location', 'Main_info'], axis=1, inplace=True)

Storing the cleaned file in CSV format.

In [118]:
df_explode_norepeat.to_csv("/Users/rahulnagial/datascience/rentfasterscraping/rentfaster_scraped_data_cleaned.csv", index=False)

The preliminary cleaning of the data is complete. Next step would be to run a more thorough exploratory data analysis (EDA) on the cleaned dataset and then build machine learning models on it.

# Next Steps

The cleaned data was stored as CSV file in the end (<a href="https://github.com/RahulNagial/Regression_Analysis" target="_blank">refer "rentfaster_scraped_data_cleaned.csv"</a>). This file would be used subsequently for more rigourous exploratory data analysis (EDA) (<a href="https://github.com/RahulNagial/Regression_Analysis" target="_blank">refer .ipynb file titled "Exploratory_Data_Analysis_and_Data_Cleaning"</a>). Insights gained from the EDA would be used to clean outliers and select features for model bulding. For more on model building, (<a href="https://github.com/RahulNagial/Regression_Analysis" target="_blank">refer .ipynb file titled "Model_Building_Tuning_Evaluation"</a>)